pkg/runtime/reconciler_test.go

```diff
@@ -0,0 +1,132 @@
+	ackmocks "github.com/aws-controllers-k8s/runtime/mocks/pkg/types"
+)
+
+func TestReconcilerUpdate(t *testing.T) {
+	require := require.New(t)
+
+	ctx := context.TODO()
+	arn := ackv1alpha1.AWSResourceName("mybook-arn")
+
+	delta := ackcompare.NewDelta()
+	delta.Add("Spec.A", "val1", "val2")
+
+	objKind := &k8srtschemamocks.ObjectKind{}
+	objKind.On("GroupVersionKind").Return(
+		k8srtschema.GroupVersionKind{
+			Group:   "bookstore.services.k8s.aws",
+			Kind:    "Book",
+			Version: "v1alpha1",
+		},
+	)
+
+	desiredRTObj := &k8srtmocks.Object{}
+	desiredRTObj.On("GetObjectKind").Return(objKind)
+
+	desiredMetaObj := &metav1mocks.Object{}
+	desiredMetaObj.On("GetAnnotations").Return(
+		map[string]string{},
+	)
+	desiredMetaObj.On("GetNamespace").Return("default")
+	desiredMetaObj.On("GetName").Return("mybook")
+	desiredMetaObj.On("GetGeneration").Return(int64(1))
+
+	desired := &ackmocks.AWSResource{}
+	desired.On("MetaObject").Return(desiredMetaObj)
+	desired.On("RuntimeObject").Return(desiredRTObj)
+
+	ids := &ackmocks.AWSResourceIdentifiers{}
+	ids.On("ARN").Return(&arn)
+
+	latest := &ackmocks.AWSResource{}
+	latest.On("Identifiers").Return(ids)
+	latest.On("Conditions").Return([]*ackv1alpha1.Condition{})
+
+	rd := &ackmocks.AWSResourceDescriptor{}
+	rd.On("GroupKind").Return(
+		&metav1.GroupKind{
+			Group: "bookstore.services.k8s.aws",
+			Kind:  "fakeBook",
+		},
+	)
+	rd.On("EmptyRuntimeObject").Return(
+		&fakeBook{},
+	)
+	rd.On("Delta", desired, latest).Return(
+		delta,
+	)
+	rd.On("UpdateCRStatus", latest).Return(false, nil)
+
+	rm := &ackmocks.AWSResourceManager{}
+	rm.On("ReadOne", ctx, desired).Return(
+		latest, nil,
+	)
+	rm.On("Update", ctx, desired, latest, delta).Return(
+		latest, nil,
+	)
+
+	rmf := &ackmocks.AWSResourceManagerFactory{}
+	rmf.On("ResourceDescriptor").Return(rd)
+
+	reg := ackrt.NewRegistry()
+	reg.RegisterResourceManagerFactory(rmf)
+
+	zapOptions := ctrlrtzap.Options{
+		Development: true,
+		Level:       zapcore.InfoLevel,
+	}
+	fakeLogger := ctrlrtzap.New(ctrlrtzap.UseFlagOptions(&zapOptions))
+	cfg := ackcfg.Config{}
+	metrics := ackmetrics.NewMetrics("bookstore")
+
+	sc := &ackmocks.ServiceController{}
+
+	// TODO(jaypipes): Place the above setup into helper functions that can be
+	// re-used by future unit tests of the reconciler code paths.
+
+	// With the above mocks and below assertions, we check that if we got a
+	// non-error return from `AWSResourceManager.ReadOne()` and the
+	// `AWSResourceDescriptor.Delta()` returned a non-empty Delta, that we end
+	// up calling the AWSResourceManager.Update() call in the Reconciler.Sync()
+	// method,
+	r := ackrt.NewReconciler(sc, rmf, fakeLogger, cfg, metrics)
+
+	err := r.Sync(ctx, rm, desired)
+	require.Nil(err)
+	rm.AssertCalled(t, "ReadOne", ctx, desired)
+	rd.AssertCalled(t, "Delta", desired, latest)
+	rm.AssertCalled(t, "Update", ctx, desired, latest, delta)
```

Comment:
I really like this! In the future we can write more unit test for different reconciliation cases